packages/@lwc/template-compiler/src/shared/renderer-hooks.ts

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { invariant, TemplateErrors } from '@lwc/errors';
+import State from '../state';
+import { BaseElement, ElementDirective } from './types';
+
+/**
+ * Config representing criteria for an element match.
+ * All conditions specified must be satisfied to be considered a match.
+ */
+export interface CustomRendererElementConfig {
+    /**
+     * Tag name to use to match an element.
+     */
+    tagName: string;
+    /**
+     * Optional namespace to match an element.
+     */
+    namespace?: string;
+    /**
+     * Specify attributes that need to be matched.
+     * This field is optional.
+     * If undefined or empty, attribute matching is skipped.
+     */
+    attributes?: string[];
+}
+
+/**
+ * Config to specify which elements and directives require a customizable renderer.
+ * An element qualified if it matches the CustomRendererElementConfig OR the directives.
+ */
+export interface CustomRendererConfig {
+    /**
+     * Element matching criteria. Element much satisfy all conditions of the CustomRendererElementConfig
+     */
+    elements: CustomRendererElementConfig[];
+    /**
+     * List of lwc directives that qualify an element. An element must use at least 1
+     * directive to be considered a match
+     */
+    directives: string[];
```

Comment:
What if this array is empty? Can it be undefined?

Reply:
It cannot be undefined. If the array is empty, it means no elements can be matched based on directives. I will add a clarifying sentence.